1-Draft/RFCNNNN-ML-Based-Command-Line-Completion.md

```diff
@@ -0,0 +1,174 @@
+
+Because of potential lag concerns (especially when the machine learning model is implemented in the cloud), querying prediction results will be an asynchronous operation.  
+
+### Interfaces
+
+```
+     public sealed class CommandLinePredictionResult
+    {
+        /// <summary>
+        /// Command line text suggested by the ML model
+        /// </summary>
+        public string SuggestionText { get; }
+
+        /// <summary>
+        /// Value indicating confidence level of prediction (0-100)
+        /// </summary>
+        public int SuggestionConfidence { get; }
+
+        /// <summary>
+        /// Null instance when no prediction is made
+        /// </summary>
+        public static readonly CommandLinePredictionResult NullInstance = new CommandLinePredictionResult();
+    }
+
+    public sealed class CommandLinePredictionStatistics
+    {
+        // TBD Prediction success statistics
+    }
+
+    interface CommandLinePredictor
+    {
+        /// <summary>
+        /// Performs machine learning prediction on provided command line fragment
+        /// </summary>
+        /// <param name="input">Command line fragment</param>
+        /// <param name="cursorIndex">Cursor index</param>
+        /// <param name="OnPredictionCallback">Call back delegate to receive prediction results</param>
+        void PredictFromInput(string input, int cursorIndex, Action<Collection<CommandLinePredictionResult>> OnPredictionCallback);
+
+        /// <summary>
+        /// Submit command line text processed by PowerShell
+        /// </summary>
+        /// <param name="processedLine">Processed command line</param>
+        /// <param name="suggestedLine">Suggested command line</param>
+        /// <param name="suggestionChosen">True if the command line was based on a model suggestion chosen by the user</param>
+        void ProcessedCommandLine(string processedLine, string suggestedLine, bool suggestionChosen);
+
+        /// <summary>
+        /// Returns prediction success statistics
+        /// </summary>
+        CommandLinePredictionStatistics GetPredictionSuccessStatisticsForCurrentUser();
+    }
+```
+
+### Feature Configuration
+
+A user can opt in to this feature through the `powershell.config.json` configuration file. The feature is disabled by default.  
+
+```json
+{"Microsoft.PowerShell:CommandLineSuggestions":"Disabled"}
+```
+
+In addition a user can see the state of this feature and configure, enable or disable it via cmdlets.  
+
+```powershell
+Get-CommandLineSuggestions
+Set-CommandLineSuggestions
+Enable-CommandLineSuggestions
+Disable-CommandLineSuggestions
+```
+
+### Examples

```

Comment:
The UX I would like to get is something similar to what happens when we type in search engine -- suggestions shows up while I'm typing and gets refreshed for each character I type.

Reply:
Please provide example text and I can include it (or you can add it later).  Thanks!